certbot-apache/certbot_apache/_internal/http_01.py

```diff
@@ -167,6 +171,10 @@ def _relevant_vhosts(self):
 
         return relevant_vhosts
 
+    def _unnamed_vhosts(self) -> List[VirtualHost]:
+        """Return all VirtualHost objects with no ServerName or ServerAlias"""
+        return [vh for vh in self.configurator.vhosts if not vh.get_names()]
```

Comment:
I think we may want to make this change:
```suggestion
        """Return all VirtualHost objects with no ServerName"""
        return [vh for vh in self.configurator.vhosts if vh.name is None]
```
Building off your instructions in your initial post, if you add something like `ServerAlias example.com` to `000-default.conf`, the vhost no longer matches here but Apache still seems to be setting the `ServerName` to the FQDN hostname:
```
$ apachectl -t -D DUMP_VHOSTS   
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using bmw.cbtesting2.com. Set the 'ServerName' directive globally to suppress this message
VirtualHost configuration:
*:80                   is a NameVirtualHost
         default server bmw.cbtesting2.com (/etc/apache2/sites-enabled/000-default.conf:1)
         port 80 namevhost bmw.cbtesting2.com (/etc/apache2/sites-enabled/000-default.conf:1)
                 alias www.example.com
         port 80 namevhost bmw.cbtesting2.com (/etc/apache2/sites-enabled/bmw.cbtesting2.conf:1)
```
Before this suggested change, the challenge will fail with this setup but it passes afterwards.

This is certainly a rarer Apache configuration, but I think we may as well handle this case too. What do you think?

Reply:
You are right, very nice catch. Apache only behaves this way if any vhost's `.server_hostname` (`ServerName`) is null, not `.names` or `.wild_names` (`ServerAlias`).